test/stress/s3benchmark.cc

```diff
@@ -0,0 +1,225 @@
+#include "upload_s3.h"
+
+#include <iostream>
+#include <fstream>
+#include <string>
+#include <climits>
+#include <time.h>
+#include <cmath>
+
+
+#include "upload.h"
+#include "upload_spooler_definition.h"
+#include "hash.h"
+#include "util/posix.h"
+#include "smalloc.h"
+
+using namespace std; // NOLINT
+
+
+class S3TestScenario {
+  public:
+    S3TestScenario(
+      const std::string &config_path,
+      const std::string &temp_path,
+      int num_uploads,
+      int avg_file_size,
+      float duplicate_file_ratio);
+    void Run();
+
+  private:
+    const float relative_variance_ = 0.1;
+    std::string config_path_;
+    std::string tmp_path_;
+    int num_uploads_;
+    int avg_file_size_;
+    float duplicate_file_ratio_;
+    upload::SpoolerDefinition *spooler_definition_;
+    upload::S3Uploader *uploader_;
+    std::vector<std::string> filenames_;
+    int unique_files_ = 0;
+
+    void GeneratePayload();
+    std::string GenerateFile();
+    std::string GenerateRandomFilename();
+    void UploadFile(const std::string &filename);
+    void Cleanup();
+    float RandomNormal();
+};
+
+S3TestScenario::S3TestScenario(
+  const std::string &config_path,
+  const std::string &temp_path,
+  int num_uploads,
+  int avg_file_size,
+  float duplicate_file_ratio)
+  : config_path_(config_path)
+  , tmp_path_(temp_path)
+  , num_uploads_(num_uploads)
+  , avg_file_size_(avg_file_size)
+  , duplicate_file_ratio_(duplicate_file_ratio)
+{
+  assert(MkdirDeep(tmp_path_, 0755, true));
+
+  std::string definition_string = "S3," + tmp_path_ + "/,s3benchmark_cvmfs@" + config_path;
+  spooler_definition_ = new upload::SpoolerDefinition(
+    definition_string,
+    shash::Algorithms::kSha1
+  );
+  uploader_ = new upload::S3Uploader(*spooler_definition_);
+}
+
+void S3TestScenario::Run()
+{
+  timespec start, payload_generated, files_uploaded, files_reuploaded, files_deleted;
+  clock_gettime(CLOCK_REALTIME, &start);
+  GeneratePayload();
+  clock_gettime(CLOCK_REALTIME, &payload_generated);
+  for(const std::string &filename : filenames_) {
+    UploadFile(filename);
+  }
+  uploader_->WaitForUpload();
+  clock_gettime(CLOCK_REALTIME, &files_uploaded);
+  for(const std::string &filename : filenames_) {
+    UploadFile(filename);
+  }
+  uploader_->WaitForUpload();
+  clock_gettime(CLOCK_REALTIME, &files_reuploaded);
+  for(const std::string &filename : filenames_) {
+    uploader_->RemoveAsync(filename);
+  }
+  uploader_->WaitForUpload();
+  clock_gettime(CLOCK_REALTIME, &files_deleted);
+  double duration_payload = payload_generated.tv_sec - start.tv_sec + 0.000000001*(payload_generated.tv_nsec - start.tv_nsec);
```

Comment:
The duration calculation should become a method to avoid duplicating the formula